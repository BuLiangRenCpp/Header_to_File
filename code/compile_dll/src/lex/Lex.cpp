#ifdef _WIN32
#define LIB_EXPORT
#endif

#include "Lex.h"
#include "usage_lex.h"
#include "output.h"
#include "Excep_dev.h"
#include "Excep_syntax.h"

using namespace std;
using namespace output;


namespace htf {
    using namespace exception;
    namespace lex {
        Lex::Lex(istream& is, const path_deal::Hpath& hpath, const std::vector<std::string>& basic_types, const std::vector<std::string>& containers)
            :_ts{ is }, _buffer{ }, _hpath{ hpath }, _basic_types{ basic_types }, _containers{ containers }
        {

        }

        bool Lex::eof()
        {
            return (this->peek().empty());
        }

        void Lex::putback(const Lexer& t)
        {
			if (t.empty()) 
				throw Excep_dev("Token_stream::putback", _LINE + "don't putback a empty Token");
			_buffer.putback(t);
        }

        void Lex::ignore(const Lexer& t)
        {
			if (this->eof())  
				return;
			if (_buffer.empty()) this->peek();	// * 会将下一个token提到buffer中，因此只需要判断 buffer 即可 
			if (_buffer.get() == t) {	 // * get -> 将 eof 提到 buffer 中的 token 删除
				return;
			}
			this->ignore(t);
        }

        void Lex::ignore_between_bracket()
        {
            if (usage::is_lbracket(this->peek().kind)) {
                char token_ch = this->get().kind;
                char target_ch = usage::ret_rbracket(token_ch);
                if (this->peek().kind == target_ch) {       // * 避免 _buffer.peek 中就是 右括号
                    this->get();
                    return;
                }
                _buffer.clear();
                _ts.putback( stream::Token{ token_ch } );
                _ts.ignore_between_bracket();
            }
        }

        Lexer Lex::get()
        {
            if (!_buffer.empty()) {
                return _buffer.get();
            }
            // ************ 忽略预处理指令、类型别名、template **************
            if (_ts.peek().kind == '#') {
                _ts.ignore_line();
                return this->get();
            }
            else if (_ts.peek().val == "using" || _ts.peek().val == "typedef") {
                _deal_define_type();
                return this->get();
            }
            else if (_ts.peek().val == "template") {
                _deal_template();
                return this->get();
            }
            // *************************************************************

            // **** const 处理：如果其后跟 type 则为 type，否则为 identifier ****
            auto token = _ts.get();
            if (token.val == "inline") return this->get();
            if (token.val == "operator") {      // * operator 重载运算符函数名
                token.val += _get_operator().val;
                return Lexer{ Lexer_kind::IDENTIFIER_KIND, token };
            }
            if (token.empty()) return Lexer{};

            Lexer lexer{ token };
            Lexer lexer_namespace;
            bool is_const_type = false;
            if (token.val == "const") {
                lexer_namespace = _get_namespace();     // 可能后面是namespace 
                if (lexer_namespace.empty() && !_is_type(_ts.peek())) return lexer;
                else lexer.val += " " + lexer_namespace.val;
                is_const_type = true;
            } else {
                _ts.putback(token);
                lexer = Lexer{};
            }        
            
            // * 函数定义语句一定以标识符开头            
            // * 作用域开头的看做 type
            if (!is_const_type) lexer_namespace = _get_namespace();             
            stream::Token peek = _ts.peek();
            if (!_is_type(peek)) {    // * 1. not type
                if (!lexer_namespace.empty()) throw Excep_syntax{_hpath.str(), _ts.line(), "in namespace" + 
                    mark_string(lexer_namespace.val) + "don't be declared type" + 
                        mark_string(peek.val) + "." + mark_string("(if not syntax-error, it's developer's error)")};
                if (peek.val == "class" || peek.val == "struct") {
                    peek = _ts.get();
                    _deal_class(false);
                    return peek;
                }
                return Lexer{ _ts.get() };
            } else {    // * 2. type  (可能以作用域开头)
                if (!is_const_type) lexer.val += lexer_namespace.val;
                lexer.kind = Lexer_kind::TYPE_KIND;     // 避免为 empty
                // 1. 多个 basic_type
                if (_is_basic_type(peek)) 
                    lexer.val += _get_basic_type().val;
                else 
                    lexer.val += _get_container().val;
                return lexer;
            }
            return Lexer{};
        }

		Lexer Lex::peek()
		{
			if (_buffer.empty_peek()) 
                _buffer.push_peek(this->get());
			return _buffer.peek();
		}

        line_t Lex::line() const
        {
            return _ts.line();
        }

        vector<string> Lex::basic_types() const
        {
            return _basic_types;
        }

        vector<string> Lex::containers() const
        {
            return _containers;
        }

        path_deal::Hpath Lex::hpath() const
        {
            return _hpath;
        }

        bool Lex::_is_basic_type(const stream::Token& t) const
        {
            return t.kind == Lexer_kind::IDENTIFIER_KIND && 
                (usage::is_basic_type_key(t.val) || my_std::is_in(t.val, _basic_types));
        }

        bool Lex::_is_container(const stream::Token& t) const
        {
            return t.kind == Lexer_kind::IDENTIFIER_KIND && 
                (usage::is_container_key(t.val) || my_std::is_in(t.val, _containers));
        }

        bool Lex::_is_type(const stream::Token& t) const
        {
            return _is_basic_type(t) || _is_container(t);
        }

        Lexer Lex::_get_namespace()
        {
            if (_ts.peek().kind != Lexer_kind::IDENTIFIER_KIND) return Lexer{};
            else {
                stream::Token token = _ts.get();
                Lexer lexer;
                while (_ts.peek().kind == Lexer_kind::DCHAR_KIND && _ts.peek().val == "::") {
                    lexer.kind = Lexer_kind::TYPE_KIND;
                    lexer.val += token.val;
                    lexer.val += _ts.get().val;
                    token = _ts.get();
                    if (_is_type(token) || token.kind != Lexer_kind::IDENTIFIER_KIND) 
                        break;
                }
                if (token.empty()) 
                    throw Excep_syntax{_hpath.str(), _ts.line(), "namespace" + mark_string(lexer.val) + "lack of type"};
                _ts.putback(token);
                return lexer;
            }
            return Lexer{};
        }

        Lexer Lex::_get_basic_type()
        {
            stream::Token token = _ts.get();
            Lexer lexer;
            if (_is_basic_type(token)) {
                lexer.kind = Lexer_kind::TYPE_KIND;
                lexer.val += token.val;
                // ******** 多个basic_type、指针、引用、数组 ***********      
                stack<char> stk;
                while (!_ts.eof()) {
                    token = _ts.get();
                    if (_is_basic_type(token)) 
                        lexer.val += " " + token.val;
                    else if (token.kind == '&' || token.kind == '*') 
                        lexer.val += token.val;
                    else if (token.kind == '[') {
                        stk.push('[');
                        lexer.val += token.val;
                    }
                    else if (token.kind == ']') {
                        if (stk.empty()) 
                            throw Excep_syntax{_hpath.str(), _ts.line(), "after array" + mark_string(lexer.val) + 
                                "lack of" + mark_char('[')};
                        stk.pop();
                        lexer.val += token.val;
                    }
                    else {
                        _ts.putback(token);
                        break;
                    }
                }
                if (!stk.empty()) 
                    throw Excep_syntax{_hpath.str(), _ts.line(), "after array" + mark_string(lexer.val) + 
                                "lack of" + mark_char(']')};
            } 
            else _ts.putback(token);
            return lexer; 
        }

        Lexer Lex::_get_container()
        {
            stream::Token token = _ts.get();
            Lexer lexer;
            if (_is_container(token)) {
                lexer.kind = Lexer_kind::TYPE_KIND;
                lexer.val += token.val;
                if (_ts.peek().kind != '<') 
                    throw Excep_syntax{_hpath.str(), _ts.line(), "after template" + mark_string(token.val) 
                        + "lack of" + mark_char('<')};

                token = _ts.get();
                stack<char> stk;    // 用于判断 <> 是否合法 
                stk.push(token.kind);
                lexer.val += token.val;
                // 1. 读取 <> 之间的部分
                while (!_ts.eof() && !stk.empty()) {
                    token = _ts.get();
                    switch (token.kind) {
                    case Lexer_kind::IDENTIFIER_KIND:
                    {
                        _ts.putback(token);
                        Lexer lexer_namespace = _get_namespace();       // * 注意可能有 namespace 
                        if (!lexer_namespace.empty()) token = _ts.peek();
                        if (_is_container(token)) 
                            lexer.val += lexer_namespace.val + _get_container().val;
                        else if (_is_basic_type(token)) 
                            lexer.val += lexer_namespace.val + _get_basic_type().val;
                        else {
                            if (!stk.empty())
                                throw Excep_syntax{_hpath.str(), _ts.line(), "in container" + mark_string(lexer.val + ">") 
                                    + ", type" + mark_string(token.val) + "don't be declared"};
                        }
                        break;
                    }
                    case '<': 
                    {
                        stk.push('<');
                        lexer.val += token.val;
                        break;
                    }
                    case '>': 
                    {
                        if (stk.empty()) 
                            throw Excep_syntax{_hpath.str(), _ts.line(), "container" + 
                                mark_string(lexer.val) + "lack of" + mark_char('>')};
                        stk.pop();
                        lexer.val += token.val;
                        break;
                    }
                    case ',':
                    {
                        if (stk.empty()) {  // * < > 之内才允许 ','
                            _ts.putback(token);
                            return lexer;
                        }
                        lexer.val += token.val + " ";
                        break;
                    }
                    default:
                    {
                        _ts.putback(token);
                        return lexer;
                    }
                    }
                }
            } 
            else if (!token.empty()) _ts.putback(token);

            // 2. 读取 <> 之后的部分 (可能有 & *)
            while (_ts.peek().kind == '&' || _ts.peek().kind == '*') 
                lexer.val += _ts.get().val;
            return lexer;          
        }

        void Lex::_deal_define_type()
        {
            stream::Token token = _ts.get();
            if (token.val == "typedef") {       // typedef  old_type  new_type ;
                if (usage::is_class_key(_ts.peek().val)) {  // typedef struct { } new_type ;
                    _ts.ignore_until();
                    _ts.ignore_between_bracket();
                }
                while (!_ts.eof()) {
                    if (_ts.peek().kind == ';') {
                        if (!my_std::is_in(token.val, _basic_types)) _basic_types.emplace_back(token.val);
                        _ts.get();
                        break;
                    }
                    token = _ts.get();
                }
            }
            else if (token.val == "using") {
                if (_ts.peek().val == "namespace")    // using namespace xx;
                    _ts.ignore();     
                else {
                    token = _ts.get();
                    if (!token.empty() && token.kind == stream::Token_kind::IDENTIFIER_KIND 
                        && _ts.peek() == '=')    // using new_type = type; 
                            if (!my_std::is_in(token.val, _basic_types)) _basic_types.emplace_back(token.val);
                    _ts.ignore();
                }
            }
            else if (!token.empty()) _ts.putback(token);
        }

        void Lex::_deal_template()
        {
            if (_ts.peek().val == "template") {
                stream::Token token = _ts.get();
                if (_ts.peek().kind != '<') 
                    throw Excep_syntax{_hpath.str(), _ts.line(), "after" + mark_string("template") + 
                        "lack of" + mark_char('<')};
                _ts.ignore_between_bracket();
                if (usage::is_class_key(_ts.peek().val)) {      // 1. 模板类
                    _deal_class();
                    _ts.ignore_between_bracket();
                    _ts.ignore();
                } else {        // 2. 模板函数  
                    _ts.ignore(stream::Token{ ')' });       // * 避免是定义语句
                    if (_ts.peek().kind == ';') _ts.ignore();
                    else _ts.ignore_between_bracket();
                }
            }   
        }

        void Lex::_deal_class(bool is_template)
        {
            stream::Token token = _ts.get();
            while (!_ts.eof()) { // class (struct) A {  }; | class DLL_API A : public AA { }; | class A;
                if (_ts.peek().kind == ':' || _ts.peek().kind == '{' || _ts.peek().kind == ';') {
                    if (is_template) {
                        if (!my_std::is_in(token.val, _containers)) _containers.emplace_back(token.val);
                    } else {
                        if (!my_std::is_in(token.val, _basic_types)) _basic_types.emplace_back(token.val);
                    } 
                    break;
                }
                token = _ts.get();
            }
            // 忽略掉 |: xxx|  { }
            if (!_ts.eof() && _ts.peek().kind != ';') {   // 避免 class A ;
                auto lbra = stream::Token{ '{' };
                _ts.ignore_until(lbra);
            }
            if (is_template == false) _ts.putback(token);
        }

        stream::Token Lex::_get_operator()
        {
            stream::Token res{};
            auto peek = _ts.peek();
            // 1. new delete
            if (peek.val == "new" || peek.val == "delete") {    
                res.val += " " + _ts.get().val;
                if (_ts.peek().kind == '[') {
                    res.val += _ts.get().val;
                    if (_ts.peek().kind == ']') res.val += _ts.get().val;
                    else throw Excep_syntax{_hpath.str(), _ts.line(), "after" + mark_string(res.val) + 
                        "lack of" + mark_char(']')};
                }
            }
            // 2. () []
            else if (usage::is_lbracket(peek.kind) && peek.kind != '<' && peek.kind != '{') {
                res.val += _ts.get().val;
                char c = usage::ret_rbracket(res.val[0]);
                if (c == _ts.peek().kind) res.val +=  _ts.get().val;
                else if (usage::is_bracket(_ts.peek().kind))
                    throw Excep_syntax{_hpath.str(), _ts.line(), "after" + mark_string(res.val) + "lack of" + 
                        mark_char(c)};
                else throw Excep_syntax{_hpath.str(), _ts.line(), "after" + mark_string("operator") + 
                    "lack of operator-char"};
            }
            // 3.  << 、 >>   ->
            else if (peek.kind == Lexer_kind::DCHAR_KIND) {
                if (peek.val == "<<" || peek.val == ">>" || peek.val == "->") res.val += _ts.get().val;
                else throw Excep_syntax{_hpath.str(), _ts.line(), mark_string(peek.val) + "isn't legal operator"};
            }
            // 4. other
            else if (usage::is_spe_ch(peek.kind)) {
                while (!_ts.eof()) {
                    if (_ts.peek().kind == '(') break;
                    res.val += _ts.get().val;
                }
            }
            else throw Excep_syntax{_hpath.str(), _ts.line(), "after" + mark_string("operator") + 
                    "lack of operator-char"};
            return res;
        }
    }
}