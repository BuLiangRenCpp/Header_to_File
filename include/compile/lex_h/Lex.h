#pragma once
#ifdef _WIN32
#define LIB_EXPORT
#endif

#include "dll.h"
#include "Lexer.h"
#include "Hpath.h"
#include "Token_stream.h"


/**
 * * Lex 对 只是 include 的文件进行 读取 自定义类型，而对于 source 则是利用之前的 
 * *     include 的信息输出 Lexer 供 Parse 使用
 * ! 1. 只返回对 parse 由于的 Lexer (函数定义语句相关的部分)
 * ! 2. 忽略 函数参数的默认值 
 * ! 3. 作用域后的标识符统一认为是 type，因此如果是方法，那么可能会报错 (这种情况一般只会出现在 {} 中，
 * !        此时是定义语句，直接忽略，故不影响程序正确执行)
 * ! 4. 对于模版直接忽略 (因为 模板一般直接定义在头文件)
 * ! 5. * inline 直接忽略
*/

namespace htf {
    namespace lex {
        class DLL_API Lex {
        public:
            Lex() = delete;
            // hpath 用来标明 is 的文件路径，用于 throw Excep_syntax
            Lex(std::istream& is, const path_deal::Hpath& hpath, const std::vector<std::string>& basic_types = {}, const std::vector<std::string>& containers = {});

			bool eof();
            // 默认放入 ';'
			void putback(const Lexer& lexer = Lexer{ stream::Token{';'} });
            // 默认忽略一条语句
			void ignore(const Lexer& lexer = Lexer{ stream::Token{ ';' } });
            void ignore_between_bracket();
			// 返回 empty Token 说明没有内容了
			Lexer get();
			Lexer peek();
            line_t line() const;

            std::vector<std::string> basic_types() const;
            std::vector<std::string> containers() const;
            path_deal::Hpath hpath() const;

        private:
            stream::Buffer<Lexer> _buffer;
            stream::Token_stream _ts;
            path_deal::Hpath _hpath;
            std::vector<std::string> _basic_types;
            std::vector<std::string> _containers;
            
            bool _is_basic_type(const stream::Token&) const;
            bool _is_container(const stream::Token&) const;
            bool _is_type(const stream::Token&) const;
            // 默认读取的是 type 之前的 namespace(即 A::A::)
            Lexer _get_namespace(bool is_type = true);
            Lexer _get_basic_type();
            Lexer _get_container();
            // 处理 'typedef B A;'  and 'using A = B;' (using namespace A; 直接忽略)
            void _deal_define_type();       
            // 处理模版：
            // 模板函数：直接忽略
            // 模板类：除了将 自定义类 加入 _containers，其余直接忽略
            // ! 如果是函数定义语句，会造成错误，因为忽略的是 { } 之间
            void _deal_template();
            // 处理 class (struct) [1 xxx] A [2 : xxx] [3 {}];  忽略 1 2
            // 默认模板类
            void _deal_class(bool is_template = true);
            stream::Token _get_operator();
        };
    }
}

