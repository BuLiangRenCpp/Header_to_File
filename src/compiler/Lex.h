#pragma once

// * 词法分析:
// * - 处理 PreProcess 产生的文件:
// *    · 对于 源文件 (-i 参数指定的) 输出 Lexer 给 Parse 处理
// *    · 对于 导入文件 只在内部记录文件定义的 type，不输出 Lexer
// * - 提取出 type:   (namespace:: 后的 id 视为 type -- std::cout 不会出现在头文件中 ->
// 因为函数定义部分会被忽略)
// * - lexer error

#include "Lexer.h"
#include "TokenStream.h"
#include "TypeTable.h"

namespace htf
{

class Lex
{
public:
    Lex() = default;
    // - tmp_file: 指定打开的文件 (必须是 Preprocess 文件)
    // - sources: 指定需要读取的源文件 (不是临时文件)，不能为空
    Lex(const path::Path& tmp_file, const std::set<FS::path>& sources);
    ~Lex() = default;

public:
    // 打开 PreProcess 的临时文件
    // - _types 不会丢失
    void open(const path::Path& tmp_file, const std::set<FS::path>& sources);

    bool        eof() const { return _eof; }
    FS::path    file() const { return _file; }
    line_t      line() const { return _ts.line(); }
    col_t       col() const { return _ts.col(); }
    std::string errors() const
    {
        std::ostringstream oss;
        for (int i = 0; i < _errors.size(); ++i) {
            oss << "[" << i << "] " << _errors[i].str() << "\n";
        }
        auto tmp = _ts.errors();
        return (_errors.empty()) ? tmp : tmp + "[lexer error]:\n" + oss.str();
    }

private:
    // 辅助 get: 主要实现识别 const type
    Lexer _get();
    Lexer _peek();

public:
    // 仅输出对 声明有用的 lexer
    Lexer get();
    Lexer peek();
    void  ignore(const Lexer& lexer = Lexer{';'});
    void  putback(const Lexer& lexer);
    void  ignore_until(const Lexer& lexer = Lexer{'{'});
    void  ignore_between_bracket();
    // 忽略 count 个 left brecket 对应的 right
    void ignore_brackets(char left_bracket, int count = 1);
    void ignore_statement();
    void ignore_current_file();
    void add_type(const std::string& type);
    void add_error(const Lexer& where, const std::string& what);

private:
    TokenStream              _ts;
    bool                     _eof     = true;
    FS::path                 _file    = "";   // 当前正在处理的源文件
    std::set<FS::path>    _sources = {};
    Buffer<Lexer>            _buffer;
    std::vector<CompilerError> _errors;
    TypeTable                _types;

private:
    Buffer<Lexer> _buffer_tmp;

private:
    // 处理 PreProcess 文件中非源文件部分
    // - 记录文件中的 type
    void update_type();
    void deal_template();
    // class and struct
    // * 目前不处理 内部类
    // * 仅将 该类名 加入 type
    // flag: 是否忽略 {} 之后的 ;
    void deal_class(bool flag = true);

private:
    // flag: 是否忽略 {} 之后的 ;
    void deal_enum(bool flag = true);
    // flag: 是否忽略 {} 之后的 ;
    void deal_union(bool flag = true);
    // 处理别名:
    // - using id = [typename] type;
    // - typedef type..  id;
    void deal_using();

private:
    // 处理基本类型，并且处理 type 之后的 *, &
    Lexer get_basic_type();
    // 同 get_basic_type，但不处理 unsigned signed 开头的 type
    std::string get_basic_type_un();
    // 处理 *, &, *const, &const
    std::string get_spe_type();

    // 处理 type 后的  <>, *, &，以及 内部类 ::type
    // * 由于语法过于复杂，因此将 <> 中的全部视为类型
    // * 对于类，还需要处理 内部类 (A<int>::value_type)
    std::string get_complex_type();
    // 读取 <> 之间
    void get_template_type(std::string& res, std::stack<char> stk);
    // 判断有它内部自己判断，调用者只需要判断 第一个是否是 id
    // 函数内部处理 ::type
    std::string get_namespace_type();
    // 处理 operator->  (视为 id -> 函数名)
    std::string get_operator_name();
};

}   // namespace htf