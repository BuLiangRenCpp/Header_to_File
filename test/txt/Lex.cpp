#include "D:\code\git\Header_to_File\src\compiler\Lex.h"


namespace htf
{

Lex::Lex(const path::Cfile& tmp_file, const std::set<std::string>& sources)
	: _ts{  }, _eof{  }, _file{  }, _sources{  }, _buffer{  }, _errors{  }, _types{  }, _buffer_tmp{  }
{

}

void Lex::open(const path::Cfile& tmp_file, const std::set<std::string>& sources)
{

}

Lexer Lex::_get()
{

}

Lexer Lex::_peek()
{

}

Lexer Lex::get()
{

}

Lexer Lex::peek()
{

}

void Lex::ignore(const Lexer& lexer)
{

}

void Lex::putback(const Lexer& lexer)
{

}

void Lex::ignore_until(const Lexer& lexer)
{

}

void Lex::ignore_between_bracket()
{

}

void Lex::ignore_brackets(char left_bracket, int count)
{

}

void Lex::ignore_statement()
{

}

void Lex::ignore_current_file()
{

}

void Lex::add_type(const std::string& type)
{

}

void Lex::add_error(const Lexer& where, const std::string& what)
{

}

void Lex::update_type()
{

}

void Lex::deal_template()
{

}

void Lex::deal_class(bool flag)
{

}

void Lex::deal_enum(bool flag)
{

}

void Lex::deal_union(bool flag)
{

}

void Lex::deal_using()
{

}

Lexer Lex::get_basic_type()
{

}

std::string Lex::get_basic_type_un()
{

}

std::string Lex::get_spe_type()
{

}

std::string Lex::get_complex_type()
{

}

void Lex::get_template_type(std::string& res, std::stack<char> stk)
{

}

std::string Lex::get_namespace_type()
{

}

std::string Lex::get_operator_name()
{

}

}   // namespace htf
