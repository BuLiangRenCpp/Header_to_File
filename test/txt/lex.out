#D:\code\git\Header_to_File\test\txt\pre.i
   5:  0:     type   :  vector<vector<int>::size_type>
   5: 31:     id     :  p
   5: 33:     schar  :  ;
   6:  1:     type   :  unsigned short
   6: 15:     schar  :  ,
   7:  1:     type   :  unsigned
   7:  9:     schar  :  ,
   8:  0:     type   :  signed
   8:  7:     schar  :  ,
   9:  0:     type   :  std::filesystem::path* const&
   9: 30:     schar  :  ,
  10:  0:     type   :  map<int>::size_type
  10: 20:     id     :  t
  10: 22:     schar  :  ;
  11:  0:     type   :   short
  11:  6:     id     :  i
  11:  9:     schar  :  =
  11: 10:     literal:  0
  11: 12:     schar  :  ;
  12:  1:     type   :   short int
  12: 11:     id     :  l
  12: 13:     key    :  const
  12: 20:     schar  :  ;
  13:  0:     type   :  const int
  13: 10:     id     :  nu
  13: 14:     schar  :  =
  13: 15:     literal:  0
  13: 17:     schar  :  ;
  14:  0:     type   :  const long
  14: 17:     schar  :  ,
  14: 18:     type   :   long int
  14: 27:     schar  :  ,
  14: 28:     key    :  const
  14: 34:     type   :   long long
  14: 44:     schar  :  ,
  14: 45:     type   :  const long long int
  14: 65:     schar  :  ,
  14: 66:     type   :  const long double
  14: 97:     schar  :  ,
  16:  1:     type   :  const Lex*
  16: 12:     id     :  l
  16: 14:     schar  :  ;
  17:  1:     type   :  LexerKind
  17: 10:     id     :  kind
  17: 15:     schar  :  ;
  18:  0:     type   :  Option*
  18:  8:     id     :  option
  18: 15:     schar  :  ;
  19:  0:     type   :  OptionBase&
  19: 12:     id     :  o
  19: 14:     schar  :  ;
  20:  0:     id     :  Ofsteam
  20:  8:     id     :  ofs
  20: 12:     schar  :  ;
  21:  0:     id     :  Ifstream
  21:  9:     id     :  ifs
  21: 13:     schar  :  ;
